src/sage/graphs/graph_decompositions/tree_decomposition.pyx

```diff
@@ -768,12 +758,186 @@ def treewidth(g, k=None, kmin=None, certificate=False, algorithm=None):
     G.add_edges(((Set(x), Set(y)) for x, y in TD), loops=False)
 
     # The Tree-Decomposition may contain a lot of useless nodes.
-    # We merge all edges between two sets S,S' where S is a subset of S'
+    # We merge all edges between two sets S, S' where S is a subset of S'
     G = reduced_tree_decomposition(G)
 
     G.name("Tree decomposition")
     return G
 
+def make_nice_tree_decomposition(T):
+    r"""
+    Return a *nice* tree decomposition (TD) of the TD `T`.
+
+    A *nice* TD `NT` is a rooted tree with four types of nodes:
+
+    - *Leaf* nodes have no children and bag size 1;
+    - *Introduce* nodes have one child: If `v \in NT` is an introduce node and
+      `w \in NT` its child, then `Bag(v) = Bag(w) \cup { x }`, where `x` is the
+      introduced node;
+    - *Forget* nodes have one child: If `v \in NT` is a forget node and
+      `w \in NT` its child, then `Bag(v) = Bag(w) \setminus { x }`, where `x` is the
+      forgotten node;
+    - *Join* nodes have two children, both identical to the parent.
+
+    INPUT::
+
+    - ``T`` -- a tree decomposition
+
+    OUTPUT::
+
+    A nice tree decomposition.
+
+    .. WARNING::
+
+        This method assumes that the vertices of the input tree `T` are hashable
+        and have attribute ``issuperset``, e.g., ``frozenset`` or
+        :class:`~sage.sets.set.Set_object_enumerated_with_category`.
+
+    TESTS::
+
+    TODO
+
+    EXAMPLES::
+
+    TODO
+    """
+    from sage.graphs.graph import Graph
+    if not isinstance(T, Graph):
```

Comment:
This test is not enough to ensure that T is a valid tree decomposition. May be the method should take both a graph and a tree decomposition as input and first call method `is_valid_tree_decomposition` ?